src/main/java/org/jenkinsci/plugins/gitclient/GitCredentialBindings.java

```diff
@@ -29,43 +29,47 @@
     default String gitToolName(Run<?, ?> run, TaskListener listener) throws IOException, InterruptedException {
 
         Executor buildExecutor = run.getExecutor();
-        ToolDescriptor<? extends GitTool> toolType = null;
+        ToolDescriptor<? extends GitTool> toolType;
         if (buildExecutor != null) {
             Node currentNode = buildExecutor.getOwner().getNode();
-            assert currentNode != null;
-            //It's on the master node
-            if (buildExecutor.getOwner().getNode().getNodeName().equals("")) {
-                toolType = Jenkins.get().getDescriptorByType(GitTool.DescriptorImpl.class);
-            } else {
-                //It's on an agent type
-                ToolLocationNodeProperty nodeToolLocation = currentNode.getNodeProperty(ToolLocationNodeProperty.class);
-                assert nodeToolLocation != null;
-                //Get Tool locations(could be many)
-                List<ToolLocationNodeProperty.ToolLocation> toolList = nodeToolLocation.getLocations();
-                //Iteration
-                for (ToolLocationNodeProperty.ToolLocation tool : toolList) {
-                    //If 'Default' Git Tool is chosen on an agent then all the Git Tools installations added
-                    // to the master nodes will be available on the agent as well
-                    //Check not an instance of jgit or jgitapache
-                    if (tool.getType().clazz.equals(GitTool.class)) {
-                        //Check for properties
-                        GitTool gitTool = new GitTool(tool.getName(), tool.getHome(), null);
-                        //Comparing and changing git home for git tool
-                        gitTool = gitTool.forNode(currentNode, listener);
-                        return gitTool.getGitExe();
+            //Check node is not null
+            if (currentNode != null) {
+                if (currentNode.getNodeName().equals("")) {
+                    //It's on the master node
```

Comment:
Thanks for the correction Justin,
The logic for getting a git tool in an agent node, depends on the `ToolLocationNodeProperty` property of that node, so if converting a Jenkins instance into a Node type, we will be depending on the `ToolLocationNodeProperty` to get a specific git tool. This approach could be less promising in case of a Jenkins controller.

Their are two ways to add tools in a Jenkins controller
* `Global Tool Configuration` - If the user depends on the tools from `Global Tool Configuration`  then the above logic will not work as Jenkins controller is using global tool and no specific tools are set if taking controller as an agent.
* `Configure System`- If the user sets tool location for the master agent and does not depends on tools from `Global Tool Configuration` then the above logic will work.

So its on the users, how they configure, if they add git tools using  `Configure System` an no git tools are added in the `Global Tool Configuration`, then any new agent created won't be getting a git tools as well.

I have made some more changes based on what I have discussed above. Thanks @justinharringa for bring this up.

Reply:
This is interesting, what concerns me is the fact that the multi branch project uses the `GitUtils.resolveGitTool` approach to resolve the correct git installation.
If I understand it correctly, it manages both Jenkins controller/agent situation. Certainly, I need to look more, in the meantime do point out any mistake in my assumptions.